### PR TITLE
[clang] Implement CTAD for type alias template.

### LINKED ISSUE #54051: Clang C++20 Feature: P1814R0 - Wording for Class Template Argument Deduction for Alias Templates
---
id: P1814R0
paper: https://wg21.link/p1814r0

### PR DESCRIPTION
**Pull request by hokein**
Fixes #54051

This patch implements the C++20 feature -- CTAD for alias templates (P1814R0, specified in https://eel.is/c++draft/over.match.class.deduct#3). It is an initial patch:
- it cover major pieces, thus it works for most cases;
- the big missing piece is to implement the associated constraints (over.match.class.deduct#3.3) for the synthesized deduction guides, see the FIXME in code and tests;
- Some enhancements on the TreeTransform&TemplateInstantiator to allow performing instantiation on `BuildingDeductionGuides` mode;

### CODE REVIEW

**Comment by cor3ntin** [comment]
I'm thrilled to see work in this area. Thanks!

Note that Clang 18 is going to branch soon, so this PR is going to target Clang 19.
However, because it took a while to stabilize other CTAD related works, i think it might be worth it to try to complete that work early in the Clang 19 cycle, so that we have a few months to weed out the bugs.

I do believe this is the last unimplemented features for our C++20 support (outside of bugs and modules)

**Comment by hokein** [comment]
> Note that Clang 18 is going to branch soon, so this PR is going to target Clang 19.

Yes, this patch will not be ready before the 18 release cut, so targeting clang 19 sounds like a good plan.

As you may notice, the implementation is incomplete, and is likely far from perfect, I can see two alternatives to proceed:
1) we keep improving it here, and landing it when it is in good/perfect shape;
2) find a way to land a relatively-good initial version, and continue to develop it incrementally;

Re 1), it is hard to predict the timeline, I'd like to finish it soon, but I also have other tasks with higher priority.
2) seems like a good option, we can hide this experimental feature with a new `-cc1` flag, when this feature is ready, we enable it by default. It is probably not a common practice for a narrow feature like this. 

Any thoughts?  @cor3ntin, @shafik, @AaronBallman

**Comment by cor3ntin** [comment]
@hokein @sam-mccall to repeat what was said in the language group last week.

We do not think a feature flag is a good fit. 

 - Until clang 19 ships (in 6 months) we do not need to stabilize the feature (but ofc we should avoid regressions, which we would not find if hidden behind a flag), and the flag does not guarantee existing features are not impacted.
 
And if we break trunk, we revert, you get extra test cases... it's perfectly fine

 - The feature should be behind a -std=C++20 flag.
 - We should not update __cpp_deduction_guides until the feature is battle tested

We do that for about every feature that is not concept/module scale (and then again, these things were tses), and this isn't big enough to warrant a novel approach.
  
I still think we should land this soon and we can start a review whenever you are ready.
If you want to land it in an incomplete state, we just need to make sure we have a good collective understanding of the remaining work that would need to be done to bring the feature to completion (tests with fix me, issues, etc)

Does that make sense to you?

**Comment by hokein** [comment]
Thanks for the summary.

> We do not think a feature flag is a good fit.
> 
> * Until clang 19 ships (in 6 months) we do not need to stabilize the feature (but ofc we should avoid regressions, which we would not find if hidden behind a flag), and the flag does not guarantee existing features are not impacted.
> 

I think this works fine with clang open-source release cycles (18 release was just cut, we have 6 months from now to polish this feature). However, this is not applicable for our internal case. At Google, we have used C++20 in production, and we have our own toolchain release cycles (which stay close to upstream main). Checking in an in-development C++20 feature without a proper safeguard poses risks to users (clang crashes, incorrect compilations, and other issues).

So having a temporary flag is useful, it enables us to incrementally test and experiment it without compromising the experience for most users. 

> And if we break trunk, we revert, you get extra test cases... it's perfectly fine

We could revert the whole patch when things are broken (and it is what we usually do). But I'm not sure revert/reland a big patch back and forth is a good idea.

This also brings a general question: what level of quality should we require from experimental features in HEAD?

> We do that for about every feature that is not concept/module scale (and then again, these things were tses), and this isn't big enough to warrant a novel approach.

Yeah, I agree that this is a narrow feature. Given that this feature is very late to the party, and C++20 has been used in some productions, I think it is probably reasonable to have a flag.

> I still think we should land this soon and we can start a review whenever you are ready.
If you want to land it in an incomplete state, we just need to make sure we have a good collective understanding of the remaining work that would need to be done to bring the feature to completion (tests with fix me, issues, etc)

+1, I think the current state should be good enough and ready for review. The remaining big piece (see the FIXME) is to implement the associated constraints (over.match.class.deduct#3.3).

**Comment by hokein** [comment]
I started a discussion on discourse: https://discourse.llvm.org/t/a-temporary-flag-for-guarding-access-to-an-in-development-c-20-feature-in-clang/76597.

**Review by shafik — COMMENTED** [review]
In order to aid the discussion on the temporary flag, I think it would be helpful to have a full set of tests that show what elements won't work right with the current set of changes.

I think this will help us gauge how impactful partial support would be and maybe help guide the discussion on alternatives.

**Comment by erichkeane** [comment]
>We could revert the whole patch when things are broken (and it is what we usually do). But I'm not sure revert/reland a big patch back and forth is a good idea.

This is effectively what we did repeatedly with the deferred concepts patch and IMO, was a really nice healthy way of doing it.  If you organize your commits (much easier on github than it was on Phab), you can present JUST the changes between revisions/fixes and get a really nice revert/fix/release cycle going.

I'd encourage just enabling by default (to avoid losing the test breadth that comes along with it), and reverting if there are sufficiently 'bad' breakages.  

Note that you'd not have to revert for all 'bugs', just those that break existing code.  We can hold off on setting the Feature Test Macro until we're comfortable with it though.

File: clang/lib/Sema/SemaInit.cpp
Comment (yronglin) [inline clang/lib/Sema/SemaInit.cpp:10893]: I was wondering if  we need to separate the code that tweaks the candidate set into another function.

Comment (hokein) [reply clang/lib/Sema/SemaInit.cpp]: Yes, and this is not the proper place to construct deduction guides for alias templates (they should live in `DeclareImplicitDeductionGuides`).

**Comment by hokein** [comment]
This patch is ready for review now.

**Review by cor3ntin — COMMENTED** [review]
Looks great generally.

I did leave a few preliminary comments.
The alias ctad mechanism is not easy to get my head around so this is going to take a few rounds

Thanks!

File: clang/lib/Sema/SemaTemplateInstantiate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplateInstantiate.cpp:1490]: Good catch. Added the bailout.

**Review by hokein — COMMENTED** [review]
Thanks for the initial review.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (cor3ntin) [reply clang/lib/Sema/SemaTemplate.cpp:2767]: An `assert(false)` may make our lives easier in the future,

Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp:2761]: Done.

File: clang/lib/Sema/SemaInit.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaInit.cpp:10612]: Do we mean Desugar here, or Canonical?

File: clang/lib/Sema/SemaTemplate.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaTemplate.cpp:2321]:
```suggestion
                                /*EvaluateConstraint=*/ true);
```
Do we mean to evaluate the constraint?  Typically we don't want to evaluate the constraint unless we are checking the decl for a call, or for constraint comparisons.  So we really don't ever substitute these ever (and the substituted versions don't really end up in the AST).

File: clang/lib/Sema/SemaTemplate.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaTemplate.cpp:2719]: The cost of a lookup can be pretty bad, I think the flag on the template is worth doing ASAP.

**Review by erichkeane — COMMENTED** [review]
We probably need an entry in ReleaseNotes.rst here as well.

File: clang/lib/Sema/SemaInit.cpp
Comment (hokein) [reply clang/lib/Sema/SemaInit.cpp:10605]: We could add one for that, but I'm not sure, it seems that in general clang doesn't provide diagnostics for features that aren't available in the current `-std=` version (at least I didn't such a diagnostic).

For example, when we use C++ concepts without `-std=c++20`, clang rejects it with a diagnostic `unknown type name 'concept'`. gcc have better diagnostics (`note: 'concept' only available with '-std=c++20' or '-fconcepts'`).

I have a somewhat-relevant FIXME below, it is about to adjust `err_deduced_non_class_template_specialization_type` message to include the type-alias templates.

File: clang/lib/Sema/SemaInit.cpp
Comment (hokein) [reply clang/lib/Sema/SemaInit.cpp:10606]: oh, good catch (this part of code has been moved and rewritten several times, I probably missed to cleanup it). Done.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp]: Yes, I think so, though the function parameter name is a bit confusing.

Here, we intend to *rewrite* the `TTP->getTypeConstraint()` (update the old template occurrences with the new template parameters). Looking at the `SubstTypeConstraint` implementation, if we set it to false, it will [attach](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaTemplateInstantiate.cpp#L2886-L2890) the old constraint to `NewTTP`.

(This is a refactoring change which lifts the existing function from `ConvertConstructorToDeductionGuideTransform` for code sharing purpose).

File: clang/lib/Sema/SemaInit.cpp
Comment (hokein) [reply clang/lib/Sema/SemaInit.cpp:10611]: Here we'd like to get the underlying type with all alias stripped. I think both `Desugar` and `Canonical` can achieve the goal here, but `Desugar` seems more explicit.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp:2719]: I think the flag change is orthogonal to the current patch, and we already do this lookup for non-type-alias CTAD implementation. 

We can address it in a followup.

File: clang/lib/Sema/SemaInit.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaInit.cpp:10605]: We actually have a TON of those sorts of things: See the warning groups named like `CXXPre20Compat`.

File: clang/lib/Sema/SemaInit.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaInit.cpp:10611]: `Desugar` doesn't remove all of the typedefs/etc, and leaves sugar in pointees/references.  So this is very likely to end up being an underlying type of, for example, `SomeTypedef*`.  Canonical canonicalizes EVERYTHING.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp]: > For the most part, we WANT the 'old' constraint to attach. All the work to instantiate the constraints end up going against the original template.

I think this is not we want for this case, we're in the process of building a new `FunctionTemplate` from the original template here, specifically building the `TemplateTypeParmDecl` for the new template. Attaching the `old` constraint to it is not correct because this will make the instantiation of the new template go against the template parameter in original template, and will lead to crashes in the normal CTAD cases, e.g. 

```
template <typename A, typename T>
concept True = true;
struct concepts {
  template <class Y>
  struct B {
    template <True<Y> Z> B(Y, Z);
  };
};

concepts::B b(0, 1);
```

Comment (erichkeane) [reply clang/lib/Sema/SemaTemplate.cpp]: Hmm... but as-written it seems like it would cause problems where the original function-template isn't instantiate-able.  It seems to me that instead of substituting it, perhaps we should be 'converting' it in some way to update the references?  I'll have to think on it, though I'm @zygoloid  can come around and answer for us.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (cor3ntin) [reply clang/lib/Sema/SemaTemplate.cpp:2677]: @erichkeane

File: clang/lib/Sema/SemaTemplate.cpp
Comment (cor3ntin) [reply clang/lib/Sema/SemaTemplate.cpp:2268]: I missed that was pre existing, (but they are pretty bad names!)

File: clang/lib/Sema/SemaTemplate.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaTemplate.cpp:2677]: I might suggest we instead genericize this to not be AliasTemplate specific, and have it take a list of NamedDecls and template args.  Then call it something like, `DeclsReferencedInTemplateArgumentList`.

I also probably wouldn't have this return a list of 'ints', and just return the pointers themselves, as this little trade back and forth is a little awkward.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp]: This case is special, we're performing a template substitution with dependent template args. I think it is fine, the constraint satisfaction check is [not performed](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaTemplate.cpp#L5029) for dependent args.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp:2677]: > I might suggest we instead genericize this to not be AliasTemplate specific, and have it take a list of NamedDecls and template args. Then call it something like, DeclsReferencedInTemplateArgumentList.

We could make this function more generic, but this function is only be used for alias-template CTAD implementation, and it is unlikely to be reused for other things, so the motivation to genericize it doesn't seem strong to me (happy to do this change if you feel strong about it).

> I also probably wouldn't have this return a list of 'ints', and just return the pointers themselves, as this little trade back and forth is a little awkward.

Yeah, I considered this in the initial version, but I found returning a list of Index can simplify the code. The caller of this function needs the `Index`, if we return a list of pointers, the caller needs to do some more work to figure out their corresponding Index.

**Comment by hokein** [comment]
@cor3ntin, @erichkeane, thanks for all review comments. I'm not sure you have reviewed the whole patch. My feeling is that the core part (`DeclareImplicitDeductionGuidesForTypeAlias`) may not have received sufficient reviews. It would be great if you can take a closer look at this part.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaTemplate.cpp]: Hmm... if it is always dependent template args we're putting in that part should be fine.  At that point it becomes just a 'replace args' instead of a substitution, which should be ok.

However, are we making sure we're replacing them 1:1 (including parent templates)?  Else the depth of some of them is going to 'change' and later substitution is going to fail.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaTemplate.cpp:2677]: The generic suggestion is because it would, I believe, make this a much more readable function with very minor changes in the 'use'.  So I think there is value there.

I see what you mean about the index, I think it is alright to leave the index part.

**Comment by erichkeane** [comment]
> @cor3ntin, @erichkeane, thanks for all review comments. I'm not sure you have reviewed the whole patch. My feeling is that the core part (`DeclareImplicitDeductionGuidesForTypeAlias`) may not have received sufficient reviews. It would be great if you can take a closer look at this part.

I've looked through it a few times(including again now) and didn't see anything that stuck out as wrong to me.  I don't believe I can review it any more thoroughly.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp]: OK, that makes sense. I made it a generic function.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp]: Yeah, we're replacing them 1:1.

**Comment by hokein** [comment]
> > @cor3ntin, @erichkeane, thanks for all review comments. I'm not sure you have reviewed the whole patch. My feeling is that the core part (`DeclareImplicitDeductionGuidesForTypeAlias`) may not have received sufficient reviews. It would be great if you can take a closer look at this part.
> 
> I've looked through it a few times(including again now) and didn't see anything that stuck out as wrong to me. I don't believe I can review it any more thoroughly.

Great, thanks. I think I have addressed all comments.

File: clang/docs/ReleaseNotes.rst
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:101]:
```suggestion
  (#GH54051).
```

File: clang/lib/Sema/SemaTemplate.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaTemplate.cpp:2908]: So this is missing https://eel.is/c++draft/over.match.class.deduct#3.3
I think we need to create a Binary (and) op expression with the requires clause of F and, a literal bool expression set to whether the arguments are deducible (do we implement the deducible part?)

File: clang/lib/Sema/SemaTemplate.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaTemplate.cpp:2965]: That FIXME should probably be merged with the one above

File: clang/lib/Sema/SemaTemplateInstantiate.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaTemplateInstantiate.cpp:1478]: Iteratively ?

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:6291]:
```suggestion
            // T can be a dependent TemplateSpecializationType when
            // performing a substitution for building a deduction guide.
```

File: clang/lib/Sema/SemaInit.cpp
Comment (hokein) [reply clang/lib/Sema/SemaInit.cpp:10628]: ah, sorry, I added it and then removed later. There are some other cases which are not supported, see the testcase `test17`.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp]: I missed this one when renaming the function. What do you think about the current one `ReferenceFinder`?

File: clang/docs/ReleaseNotes.rst
Comment (hokein) [reply clang/docs/ReleaseNotes.rst]: Done.

File: clang/docs/ReleaseNotes.rst
Comment (hokein) [reply clang/docs/ReleaseNotes.rst:101]: This doesn't follow the pattern used in this file, we use `#<number> https://github.com/llvm/llvm-project/issues/<nummber>`.

File: clang/lib/Sema/SemaInit.cpp
Comment (hokein) [reply clang/lib/Sema/SemaInit.cpp:10632]: Are you suggesting to address it now? My original plan is to address it in a followup, but it seems small enough to do it in this patch.

File: clang/docs/ReleaseNotes.rst
Comment (cor3ntin) [reply clang/docs/ReleaseNotes.rst:101]: That changes recently (either way, you were missing <>)

Comment (hokein) [reply clang/docs/ReleaseNotes.rst]: I see, I were not aware of it. Switched to the `GH` form.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaTemplate.cpp]: > I missed this one when renaming the function. What do you think about the current one `ReferenceFinder`?

ReferenceFinder is a little overloaded of a name :)  Perhaps `TypeReferencedFinder`? Note the 'd' which I think prevents it from being mistaken with a "finds all things with an L or R value reference".

Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp]: OK, I'm not sure `Type` is the best fit, I'd stay close with template params.  Renamed it to `TemplateParamsReferencedFinder` (if we want to use a less-verbose name, I'm fine with `Finder` since this is a structure defined in a function body).

File: clang/docs/ReleaseNotes.rst
Comment (erichkeane) [inline clang/docs/ReleaseNotes.rst:99]:
```suggestion
- Initial support for class template argument deduction (CTAD) for type alias
```

**Review by erichkeane — COMMENTED** [review]
I'm happy when Corentin is, so he can approve when he's ready.

File: clang/lib/Sema/SemaInit.cpp
Comment (cor3ntin) [reply clang/lib/Sema/SemaInit.cpp:10632]: I think we might as well, yes (unless it's more complicated than i think it is)

File: clang/lib/Sema/SemaTemplate.cpp
Comment (cor3ntin) [reply clang/lib/Sema/SemaTemplate.cpp]: I think it's fine to leave that to a subsequent patch

**Comment by cor3ntin** [comment]
Can you mark the status of the paper in cxx_status.html as partial (with a `<details></details>` tag explaining what is missing)?
Can you update the commit message / changelog to refer to the paper number?

Thanks

File: clang/docs/ReleaseNotes.rst
Comment (h-vetinari) [inline clang/docs/ReleaseNotes.rst:100]: Probably should update to at least a `<td class="partial" align="center">Partial</td>` under
https://github.com/llvm/llvm-project/blob/e946b5a87b2db307da076093d0a9a72ecb4ec089/clang/www/cxx_status.html#L843-L847

File: clang/docs/ReleaseNotes.rst
Comment (hokein) [reply clang/docs/ReleaseNotes.rst]: Done.

File: clang/lib/Sema/SemaInit.cpp
Comment (hokein) [reply clang/lib/Sema/SemaInit.cpp]: Sure, please take a look on the new change. I also rename the diagnostic to `err_deduced_non_class_or_alias_template_specialization_type`.

File: clang/docs/ReleaseNotes.rst
Comment (hokein) [reply clang/docs/ReleaseNotes.rst:95]: Done.

**Review by Endilll — COMMENTED** [review]
`Sema.h` changes look good.

**Review by cor3ntin — APPROVED** [review]
LGTM
Thanks for working on that!

File: clang/lib/Sema/SemaInit.cpp
Comment (hokein) [reply clang/lib/Sema/SemaInit.cpp:10750]: Yes, exactly.

**Comment by hokein** [comment]
Thank you for all the reviews, I will wait 1 or 2 days before merging it in case anyone has more comments.

**Comment by cor3ntin** [comment]
> Thank you for all the reviews, I will wait 1 or 2 days before merging it in case anyone has more comments.

Sure!
Do you plan to keep working on completing this feature?

**Comment by hokein** [comment]
> > Thank you for all the reviews, I will wait 1 or 2 days before merging it in case anyone has more comments.
> 
> Sure! Do you plan to keep working on completing this feature?

Yeah, I plan to complete this feature before the next clang release. If anyone is willing to help, that would be even greater.

BTW, should we close https://github.com/llvm/llvm-project/issues/54051 with this PR? or should we keep it open until the feature is fully implemented? The `is_deducible` part is a major missing piece, I think we can close #54051 and open a separate issue to track the missing piece, WDYT?

**Comment by cor3ntin** [comment]
> BTW, should we close #54051 with this PR? or should we keep it open until the feature is fully implemented? The `is_deducible` part is a major missing piece, I think we can close #54051 and open a separate issue to track the missing piece, WDYT?

Sounds good

**Pull request merged**
